File: kernel/test/kernel-test.c
/* kernel-test — the reactive core's contract, over the C ABI. Mirrors what
 * test/static-constraint.test.mjs and the runtime's unit tests pin for
 * reactive.ts / attributes.ts. A host program: libc is fine here. */
#include "declare_kernel.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

static int fails = 0, checks = 0;
#define CHECK(c) do { checks++; if (!(c)) { fails++; printf("  ✗ %s:%d  %s\n", __FILE__, __LINE__, #c); } } while (0)

/* ── an image, by hand (the same layout kernel/image.mjs writes) ────────── */
typedef struct { unsigned char *p; uint32_t n, cap; } Buf;
static void put(Buf *b, const void *d, uint32_t n) { if (b->n + n > b->cap) { b->cap = (b->n + n) * 2 + 64; b->p = realloc(b->p, b->cap); } memcpy(b->p + b->n, d, n); b->n += n; }
static void pad8(Buf *b) { static const unsigned char z[8]; uint32_t r = b->n % 8; if (r) put(b, z, 8 - r); }
static void u32(Buf *b, uint32_t v) { put(b, &v, 4); }
static void f64(Buf *b, double v) { put(b, &v, 8); }

typedef struct { int32_t target; uint8_t kind, flags; uint32_t *edges; uint32_t nedges; uint32_t *code; uint32_t ncode; uint32_t body; } R;

static Buf image(uint32_t nelems, const uint32_t *nslots, const uint8_t *kinds, const double *init,
                 const R *rules, uint32_t nrules, const double *consts, uint32_t nconsts) {
  Buf b = { 0, 0, 0 };
  uint32_t ncells = 0, nedges = 0, ncode = 0;
  for (uint32_t i = 0; i < nelems; i++) ncells += nslots[i];
  for (uint32_t i = 0; i < nrules; i++) { nedges += rules[i].nedges; ncode += rules[i].ncode; }
  u32(&b, DK_MAGIC); u32(&b, DK_VERSION); u32(&b, nelems); u32(&b, ncells); u32(&b, nrules); u32(&b, nedges); u32(&b, ncode); u32(&b, nconsts);
  put(&b, kinds, ncells); pad8(&b);
  for (uint32_t i = 0, base = 0; i < nelems; i++) { u32(&b, base); u32(&b, 0xffffffffu); u32(&b, nslots[i]); base += nslots[i]; } pad8(&b);
  for (uint32_t i = 0, e0 = 0, c0 = 0; i < nrules; i++) {
    const R *r = &rules[i];
    put(&b, &r->target, 4); put(&b, &r->kind, 1); put(&b, &r->flags, 1); { uint16_t z = 0; put(&b, &z, 2); }
    u32(&b, e0); u32(&b, r->nedges); u32(&b, c0); u32(&b, r->ncode); u32(&b, r->body);
    e0 += r->nedges; c0 += r->ncode;
  } pad8(&b);
  for (uint32_t i = 0; i < nrules; i++) for (uint32_t e = 0; e < rules[i].nedges; e++) u32(&b, rules[i].edges[e]); pad8(&b);
  for (uint32_t i = 0; i < nrules; i++) for (uint32_t c = 0; c < rules[i].ncode; c++) u32(&b, rules[i].code[c]); pad8(&b);
  for (uint32_t i = 0; i < nconsts; i++) f64(&b, consts[i]); pad8(&b);
  for (uint32_t i = 0; i < ncells; i++) f64(&b, init ? init[i] : 0.0); pad8(&b);
  return b;
}

/* ── the host: bodies as C functions, a log of what ran ─────────────────── */
static dk_kernel *K;
static char logbuf[4096]; static int logn;
static void logf_(const char *s) { logn += snprintf(logbuf + logn, sizeof logbuf - logn, "%s ", s); }
static int after_count, change_count, after_arm, change_arm, ended;
static double body_value[64]; static int body_runs[64];
static double (*body_fn[64])(dk_kernel *k, uint32_t rule);

static double body(void *ctx, uint32_t rule, uint32_t elem, int32_t target) {
  (void)ctx; (void)elem; (void)target;
  body_runs[rule]++;
  char s[32]; snprintf(s, sizeof s, "b%u", rule); logf_(s);
  return body_fn[rule] ? body_fn[rule](K, rule) : body_value[rule];
}
static int after_steps(void *ctx) { (void)ctx; if (after_arm > 0) { after_arm--; after_count++; logf_("after"); return 1; } return 0; }
static int fire_changes(void *ctx) { (void)ctx; if (change_arm > 0) { change_arm--; change_count++; logf_("change"); return 1; } return 0; }
static void end_chain(void *ctx) { (void)ctx; ended++; }
static int last_err; static uint32_t last_bad;
static void on_error(void *ctx, int code, uint32_t rule) { (void)ctx; last_err = code; last_bad = rule; }
static const dk_host HOST = { 0, body, after_steps, fire_changes, end_chain, on_error };

static dk_kernel *load(Buf *b, dk_caps caps) {
  uint32_t need = kernel_arena_size(b->p, b->n, &caps);
  void *arena = malloc(need);
  dk_kernel *k = kernel_load(b->p, b->n, &caps, arena, need, &HOST);
  K = k; logn = 0; logbuf[0] = 0;
  memset(body_runs, 0, sizeof body_runs); memset(body_fn, 0, sizeof body_fn);
  return k;
}

/* ── scenarios ──────────────────────────────────────────────────────────── */

/* 1. a chain of EXPR rules: c1 = c0*2, c2 = c1+1 — one settle, in order, gated */
static void test_chain(void) {
  uint32_t nslots[1] = { 3 }; uint8_t kinds[3] = { 0, 0, 0 }; double init[3] = { 1, 0, 0 };
  double consts[2] = { 2, 1 };
  uint32_t e1[1] = { 0 }, code1[] = { DK_OP_LOAD, 0, DK_OP_CONST, 0, DK_OP_MUL, DK_OP_END };
  uint32_t e2[1] = { 1 }, code2[] = { DK_OP_LOAD, 1, DK_OP_CONST, 1, DK_OP_ADD, DK_OP_END };
  R rules[2] = { { 1, DK_EXPR, 0, e1, 1, code1, 6, 0 }, { 2, DK_EXPR, 0, e2, 1, code2, 6, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 2, consts, 2);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 0 });
  CHECK(k != 0);
  double *t = kernel_table(k);
  kernel_run(k, 0); kernel_run(k, 1);
  CHECK(t[1] == 2 && t[2] == 3);
  kernel_settle(k);                        /* the landings' own wakes drain, as in the JS core */
  CHECK(kernel_write(k, 0, 10) == DK_OK);
  CHECK(kernel_pending(k) == 1);
  CHECK(t[1] == 2);                       /* dependents recompute at the settle, not at the write */
  int32_t runs = kernel_settle(k);
  CHECK(runs == 2 && t[1] == 20 && t[2] == 21);
  CHECK(kernel_settle(k) == 0);           /* quiescent */
  kernel_write(k, 0, 10);                  /* equal value: gated, nothing queued */
  CHECK(kernel_pending(k) == 0 && kernel_settle(k) == 0);
  uint32_t dirty[8]; CHECK(kernel_dirty(k, dirty, 8) == 3);  /* 0, 1, 2 moved since load */
  CHECK(kernel_dirty(k, dirty, 8) == 0);
  printf("  chain ok\n");
}

/* 2. a diamond over BODY rules: d = b + c, b = a, c = a — a moves → b, c, d run once each */
static double b_read(dk_kernel *k, uint32_t r) { (void)r; return kernel_table(k)[0]; }
static double d_sum(dk_kernel *k, uint32_t r) { (void)r; return kernel_table(k)[1] + kernel_table(k)[2]; }
static void test_diamond(void) {
  uint32_t nslots[1] = { 4 }; uint8_t kinds[4] = { 0, 0, 0, 0 }; double init[4] = { 1, 0, 0, 0 };
  uint32_t ea[1] = { 0 }, ebc[2] = { 1, 2 };
  R rules[3] = { { 1, DK_BODY, 0, ea, 1, 0, 0, 0 }, { 2, DK_BODY, 0, ea, 1, 0, 0, 0 }, { 3, DK_BODY, 0, ebc, 2, 0, 0, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 3, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 0 });
  body_fn[0] = b_read; body_fn[1] = b_read; body_fn[2] = d_sum;
  kernel_run(k, 0); kernel_run(k, 1); kernel_run(k, 2);
  CHECK(kernel_table(k)[3] == 2);
  kernel_settle(k);
  memset(body_runs, 0, sizeof body_runs); logn = 0;
  kernel_write(k, 0, 5);
  int32_t runs = kernel_settle(k);
  /* FIFO with re-queue: b, c queued by a; d queued by b's apply, runs once after c (its second wake coalesces) */
  CHECK(runs == 3);
  CHECK(body_runs[0] == 1 && body_runs[1] == 1 && body_runs[2] == 1);
  CHECK(kernel_table(k)[3] == 10);
  CHECK(strcmp(logbuf, "b0 b1 b2 ") == 0);
  printf("  diamond ok (%s)\n", logbuf);
}

/* 3. phases: a phase-1 (draw) rule runs after every phase-0 rule, even when queued first */
static void test_phases(void) {
  uint32_t nslots[1] = { 3 }; uint8_t kinds[3] = { 0, 0, 1 }; double init[3] = { 0, 0, 0 };
  uint32_t e0[1] = { 0 };
  R rules[2] = { { 2, DK_BODY, DK_PHASE1, e0, 1, 0, 0, 0 }, { 1, DK_BODY, 0, e0, 1, 0, 0, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 2, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 0 });
  body_fn[1] = b_read; body_value[0] = 1;   /* the draw rule targets a REF cell: returns "changed" */
  kernel_run(k, 0); kernel_run(k, 1); kernel_settle(k);   /* land both (an unlanded rule never wakes) */
  kernel_write(k, 0, 1);
  logn = 0; kernel_settle(k);
  CHECK(strcmp(logbuf, "b1 b0 ") == 0);
  printf("  phases ok (%s)\n", logbuf);
}

/* 4. the cycle guard: a rule that re-dirties its own input */
static double cyc(dk_kernel *k, uint32_t r) { (void)r; double v = kernel_table(k)[0] + 1; kernel_set(k, 0, v); return v; }
static void test_cycle(void) {
  uint32_t nslots[1] = { 2 }; uint8_t kinds[2] = { 0, 0 }; double init[2] = { 0, 0 };
  uint32_t e0[1] = { 0 };
  R rules[1] = { { 1, DK_BODY, 0, e0, 1, 0, 0, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 1, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 0 });
  body_fn[0] = cyc; last_err = 0;
  kernel_run(k, 0);                          /* landing: its own write wakes it */
  ended = 0;
  kernel_write(k, 0, 5);                     /* (the landing left cell 0 at 1: a write of 1 would gate) */
  int32_t r = kernel_settle(k);
  CHECK(r == DK_ERR_CYCLE && last_err == DK_ERR_CYCLE && last_bad == 0);
  CHECK(ended == 1);
  CHECK(kernel_pending(k) == 0);            /* abandoned: the queue is clear */
  kernel_write(k, 0, 100);                   /* and a later write queues again */
  CHECK(kernel_pending(k) == 1);
  printf("  cycle ok\n");
}

/* 5. ownership: a bound slot refuses an author write; a yielding owner steps aside */
static void test_ownership(void) {
  uint32_t nslots[1] = { 3 }; uint8_t kinds[3] = { 0, 0, 0 }; double init[3] = { 1, 0, 0 };
  uint32_t e0[1] = { 0 };
  R rules[2] = { { 1, DK_BODY, 0, e0, 1, 0, 0, 0 }, { 2, DK_BODY, DK_YIELDING, e0, 1, 0, 0, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 2, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 0 });
  body_fn[0] = b_read; body_fn[1] = b_read;
  CHECK(kernel_own(k, 1, 0) == DK_OK && kernel_own(k, 2, 1) == DK_OK);
  CHECK(kernel_own(k, 1, 1) == DK_ERR_BOUND);          /* already bound, non-yielding */
  kernel_run(k, 0); kernel_run(k, 1);
  CHECK(kernel_write(k, 1, 7) == DK_ERR_OWNED);         /* author write to a bound slot */
  CHECK(kernel_table(k)[1] == 1);
  CHECK(kernel_write(k, 2, 7) == DK_OK);                /* the yielding derive is displaced */
  CHECK(kernel_owner(k, 2) == -1 && kernel_is_set(k, 2) == 1 && kernel_table(k)[2] == 7);
  kernel_write(k, 0, 3); kernel_settle(k);
  CHECK(kernel_table(k)[1] == 3 && kernel_table(k)[2] == 7);   /* the disposed derive never runs again */
  CHECK(kernel_set(k, 1, 9) == DK_OK && kernel_table(k)[1] == 9);   /* setBound ignores ownership */
  printf("  ownership ok\n");
}

/* 6. dynamic tracking: a body reads one of two cells depending on a third; edges follow */
static double pick(dk_kernel *k, uint32_t r) {
  (void)r; double *t = kernel_table(k);
  kernel_track(k, 0);
  if (t[0] != 0) { kernel_track(k, 1); return t[1]; }
  kernel_track(k, 2); return t[2];
}
static void test_dynamic(void) {
  uint32_t nslots[1] = { 4 }; uint8_t kinds[4] = { 0, 0, 0, 0 }; double init[4] = { 0, 10, 20, 0 };
  R rules[1] = { { 3, DK_DYNAMIC, 0, 0, 0, 0, 0, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 1, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 16 });
  body_fn[0] = pick;
  kernel_run(k, 0); CHECK(kernel_table(k)[3] == 20);
  memset(body_runs, 0, sizeof body_runs);
  kernel_write(k, 1, 11); kernel_settle(k);
  CHECK(body_runs[0] == 0);                  /* not a dependency this run */
  kernel_write(k, 2, 21); kernel_settle(k);
  CHECK(body_runs[0] == 1 && kernel_table(k)[3] == 21);
  kernel_write(k, 0, 1); kernel_settle(k);
  CHECK(body_runs[0] == 2 && kernel_table(k)[3] == 11);
  kernel_write(k, 2, 99); kernel_settle(k);
  CHECK(body_runs[0] == 2);                  /* the branch not taken is not a dependency */
  kernel_write(k, 1, 12); kernel_settle(k);
  CHECK(body_runs[0] == 3 && kernel_table(k)[3] == 12);
  CHECK(kernel_active(k) == -1);
  printf("  dynamic ok\n");
}

/* 7. the close: afterSettle steps and change events loop back; the limit trips */
static void test_close(void) {
  uint32_t nslots[1] = { 1 }; uint8_t kinds[1] = { 0 }; double init[1] = { 0 };
  Buf b = image(1, nslots, kinds, init, 0, 0, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 0, 0 });
  after_arm = 2; change_arm = 1; after_count = change_count = 0; logn = 0;
  kernel_write(k, 0, 1);
  CHECK(kernel_settle(k) >= 0);
  CHECK(after_count == 2 && change_count == 1);
  CHECK(strcmp(logbuf, "after after change ") == 0);   /* steps first, to quiescence, then the change */
  printf("  close ok (%s)\n", logbuf);
  after_arm = 1000; last_err = 0; logn = 0;
  kernel_write(k, 0, 2);
  CHECK(kernel_settle(k) == DK_ERR_AFTER && last_err == DK_ERR_AFTER);
  after_arm = 0;
}

/* 8. suspend / resume / dispose; runtime-added rules; NaN propagates */
static void test_lifecycle(void) {
  uint32_t nslots[1] = { 3 }; uint8_t kinds[3] = { 0, 0, 0 }; double init[3] = { 1, 0, 0 };
  uint32_t e0[1] = { 0 };
  R rules[1] = { { 1, DK_BODY, 0, e0, 1, 0, 0, 0 } };
  Buf b = image(1, nslots, kinds, init, rules, 1, 0, 0);
  dk_kernel *k = load(&b, (dk_caps){ 0, 0, 4, 16 });
  body_fn[0] = b_read; body_fn[1] = b_read;
  kernel_run(k, 0);
  int32_t r2 = kernel_add_rule(k, 2, DK_BODY, 0, e0, 1, 0, 0, 0);
  CHECK(r2 == 1);
  kernel_run(k, 1); CHECK(kernel_table(k)[2] == 1);
  kernel_suspend(k, 0);
  kernel_write(k, 0, 5); kernel_settle(k);
  CHECK(kernel_table(k)[1] == 1 && kernel_table(k)[2] == 5);    /* suspended: inert; the added rule woke */
  kernel_resume(k, 0); CHECK(kernel_table(k)[1] == 5);           /* resumed: re-evaluated now */
  kernel_dispose(k, 1);
  kernel_write(k, 0, 6); kernel_settle(k);
  CHECK(kernel_table(k)[1] == 6 && kernel_table(k)[2] == 5);    /* disposed never runs */
  kernel_write(k, 0, NAN); kernel_settle(k); CHECK(isnan(kernel_table(k)[1]));
  kernel_write(k, 0, NAN); CHECK(kernel_pending(k) == 1);       /* NaN !== NaN: propagates again */
  printf("  lifecycle ok\n");
}

int main(void) {
  printf("kernel-test\n");
  test_chain(); test_diamond(); test_phases(); test_cycle(); test_ownership(); test_dynamic(); test_close(); test_lifecycle();
  printf("%d checks, %d failed\n", checks, fails);
  return fails ? 1 : 0;
}

File: kernel/test/kernel.test.mjs
// kernel.test — the kernel under WebAssembly, from JS: the C scenarios again
// through image.mjs + loader.mjs, then the SETTLE TRACES: the same graph and
// the same writes driven through the JS reactive core (runtime/dist) and
// through the kernel, compared run for run. That comparison is Phase A's gate.
//
//   node kernel/test/kernel.test.mjs      (after: node kernel/build.mjs wasm)
import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import assert from "node:assert/strict";
import { buildImage, assemble, KIND, FLAG, CELL } from "../image.mjs";
import { loadKernel, ERR } from "../loader.mjs";
// THE ORACLE is the previous JS core — main's untouched runtime — not this
// tree's reactive.ts, which is the kernel itself.
import { Cell, Constraint, settle as jsSettle, afterSettle } from "/Users/temkin/Code/Declare/runtime/dist/reactive.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const WASM = readFileSync(path.join(HERE, "../build/kernel.wasm"));
let passed = 0;
const test = async (name, fn) => { try { await fn(); passed++; console.log("  ok   " + name); } catch (e) { console.log("  FAIL " + name + "\n       " + (e.stack ?? e)); process.exitCode = 1; } };

const host = (bodies, extra = {}) => ({ log: [], body(rule, elem, target) { this.log.push("b" + rule); return bodies[rule]?.(rule, elem, target) ?? 0; }, ...extra });

console.log("kernel.test (wasm)");

await test("chain of EXPR rules: gated, settled once, dirty list", async () => {
  const consts = [];
  const img = buildImage({
    elems: [{ nslots: 3 }], cells: [{ init: 1 }, {}, {}],
    rules: [
      { target: 1, kind: KIND.EXPR, edges: [0], code: assemble(consts, ["load", 0], ["const", 2], "mul") },
      { target: 2, kind: KIND.EXPR, edges: [1], code: assemble(consts, ["load", 1], ["const", 1], "add") },
    ], consts,
  });
  const k = await loadKernel(WASM, img, host({}));
  k.run(0); k.run(1); k.settle();
  assert.deepEqual(Array.from(k.table), [1, 2, 3]);
  k.write(0, 10); assert.equal(k.pending(), true); assert.equal(k.table[1], 2);
  assert.equal(k.settle(), 2); assert.deepEqual(Array.from(k.table), [10, 20, 21]);
  k.write(0, 10); assert.equal(k.pending(), false);
  assert.deepEqual(k.dirty(), [1, 2, 0]); assert.deepEqual(k.dirty(), []);
});

await test("BODY rules call back synchronously; REF targets carry the wake", async () => {
  const img = buildImage({
    elems: [{ nslots: 3 }], cells: [{ init: 4 }, { kind: CELL.REF }, {}],
    rules: [{ target: 1, kind: KIND.BODY, edges: [0], body: 7 }, { target: 2, kind: KIND.BODY, edges: [1], body: 8 }],
  });
  const refs = [null, null, null];
  const h = host({
    0: () => { const v = "n" + k.table[0]; const changed = refs[1] !== v; refs[1] = v; return changed ? 1 : 0; },
    1: () => refs[1].length,
  });
  const k = await loadKernel(WASM, img, h);
  k.run(0); k.run(1); k.settle(); assert.equal(k.table[2], 2);
  k.write(0, 123); assert.equal(k.settle(), 2); assert.equal(k.table[2], 4);
  h.log.length = 0; k.write(0, 123.0); assert.equal(k.settle(), 0);      // gated at the number
  k.write(0, 321); k.settle(); assert.deepEqual(h.log, ["b0", "b1"]);     // the REF changed ("n321" vs "n123") → b1 runs, lands the same length (gated there)
});

await test("ownership, cycle, dynamic tracking, close", async () => {
  const img = buildImage({
    elems: [{ nslots: 4 }], cells: [{ init: 0 }, { init: 10 }, { init: 20 }, {}],
    rules: [{ target: 3, kind: KIND.DYNAMIC }, { target: 1, kind: KIND.BODY, flags: FLAG.YIELDING, edges: [0] }],
  });
  let steps = 0, changes = 0, errors = [];
  const h = host({
    0: () => { k.track(0); if (k.table[0] !== 0) { k.track(1); return k.table[1]; } k.track(2); return k.table[2]; },
    1: () => k.table[0] + 100,
  }, { afterSteps: () => steps-- > 0, fireChanges: () => changes-- > 0, error: (c, r) => errors.push([c, r]) });
  const k = await loadKernel(WASM, img, h, { dyn_edges: 16 });
  k.run(0); k.settle(); assert.equal(k.table[3], 20);
  h.log.length = 0; k.write(1, 11); k.settle(); assert.deepEqual(h.log, []);          // not read this run
  k.write(2, 21); k.settle(); assert.equal(k.table[3], 21);
  k.write(0, 1); k.settle(); assert.equal(k.table[3], 11);
  assert.equal(k.active[0], -1);
  // ownership: the yielding derive owns cell 1; an author write displaces it
  assert.equal(k.own(1, 1), 0); k.run(1); k.settle(); assert.equal(k.table[1], 101);
  assert.equal(k.write(1, 5), 0); assert.equal(k.owner(1), -1); assert.equal(k.isSet(1), true);
  k.write(0, 2); k.settle(); assert.equal(k.table[1], 5);                              // never runs again
  // the close: steps then changes, each looping back
  steps = 2; changes = 1; k.write(0, 3); assert.ok(k.settle() >= 0); assert.equal(steps, -2); assert.equal(changes, -1);   // steps, steps, (none) → change → steps asked again → (none)
  steps = 1000; k.write(0, 4); assert.equal(k.settle(), ERR.AFTER); assert.deepEqual(errors.at(-1), [ERR.AFTER, -1]); steps = 0;   // no rule to blame: NONE crosses as i32 −1
});

// ── settle traces: the JS core and the kernel on one graph ──────────────────
//
// A random DAG of numeric rules — each `sum of some earlier cells, plus a
// constant` — with the same random write sequence applied to both. Compared:
// every cell's value after every settle, and the ORDER rules ran in (the
// visit policy is the JS core's: FIFO by invalidation with re-queue).
await test("settle traces: kernel ≡ reactive.ts on 60 random graphs", async () => {
  let seed = 12345;
  const rnd = (n) => { seed = (seed * 1103515245 + 12345) & 0x7fffffff; return seed % n; };
  for (let g = 0; g < 60; g++) {
    const N = 4 + rnd(12), R = 2 + rnd(10);
    const rules = [];
    for (let i = 0; i < R; i++) {
      const target = 1 + rnd(N - 1);
      const reads = new Set(); const nr = 1 + rnd(3);
      for (let j = 0; j < nr; j++) { const c = rnd(N); if (c !== target) reads.add(c); }
      if (reads.size === 0) reads.add(target === 0 ? 1 : 0);
      rules.push({ target, reads: [...reads], k: rnd(5) });
    }
    // no two rules own one cell (the one-owner rule); drop duplicates
    const seen = new Set(); const uniq = rules.filter((r) => !seen.has(r.target) && seen.add(r.target));
    // ── JS core ──
    const cells = Array.from({ length: N }, () => new Cell());
    const vals = Array.from({ length: N }, (_, i) => i);
    const jsLog = [];
    const cs = uniq.map((r, i) => new Constraint("r" + i, () => { jsLog.push(i); return r.reads.reduce((s, c) => s + vals[c], r.k); }, (v) => { if (vals[r.target] !== v) { vals[r.target] = v; cells[r.target].changed(); } }));
    const probes = uniq.map((r) => () => { for (const c of r.reads) cells[c].track(); });
    // ── kernel ── (BODY rules with static edges, the same bodies)
    const img = buildImage({ elems: [{ nslots: N }], cells: Array.from({ length: N }, (_, i) => ({ init: i })), rules: uniq.map((r) => ({ target: r.target, kind: KIND.BODY, edges: r.reads })) });
    const h = host(Object.fromEntries(uniq.map((r, i) => [i, () => r.reads.reduce((s, c) => s + k.table[c], r.k)])));
    const k = await loadKernel(WASM, img, h);
    // the same landing order, then the same drain
    let cyc = false;
    try { for (let i = 0; i < cs.length; i++) cs[i].wire(probes[i]); jsSettle(); } catch (e) { if (!/cycle/.test(String(e))) throw e; cyc = true; }
    for (let i = 0; i < uniq.length; i++) { k.run(i); k.own(uniq[i].target, i); }   // a bound slot has an owner (attributes.ts own)
    const kr = k.settle();
    if (cyc) { assert.equal(kr, ERR.CYCLE, `graph ${g}: JS cycled, kernel did not`); continue; }
    assert.ok(kr >= 0, `graph ${g}: kernel errored ${kr} where JS settled`);
    for (let step = 0; step < 12; step++) {
      jsLog.length = 0; h.log.length = 0;
      const c = rnd(N), v = rnd(50) - 10;
      // a JS author write: the cell value + wake (owned cells refused in both)
      const owned = uniq.some((r) => r.target === c);
      if (owned) { assert.equal(k.write(c, v), ERR.OWNED, `graph ${g}: owned cell accepted`); continue; }
      if (vals[c] !== v) { vals[c] = v; cells[c].changed(); }
      let jsErr = null; try { jsSettle(); } catch (e) { jsErr = e; }
      k.write(c, v); const r = k.settle();
      if (jsErr) { assert.equal(r, ERR.CYCLE, `graph ${g} step ${step}: JS threw, kernel returned ${r}`); break; }
      assert.deepEqual(Array.from(k.table), vals, `graph ${g} step ${step}: values differ`);
      assert.deepEqual(h.log.map((s) => +s.slice(1)), jsLog, `graph ${g} step ${step}: run order differs`);
    }
  }
});

await test("afterSettle ordering matches: steps run at the close, writes fold into the same settle", async () => {
  // JS: a step that writes; the settle loops. Kernel: the host's afterSteps does the same through kernel.set
  const img = buildImage({ elems: [{ nslots: 2 }], cells: [{ init: 0 }, {}], rules: [{ target: 1, kind: KIND.BODY, edges: [0] }] });
  const trace = [];
  let armed = false;
  const h = host({ 0: () => { trace.push("run"); return k.table[0] * 2; } }, { afterSteps: () => { if (!armed) return false; armed = false; trace.push("step"); k.set(0, 7); return true; } });
  const k = await loadKernel(WASM, img, h);
  k.run(0); k.settle(); trace.length = 0;
  armed = true; k.write(0, 3); k.settle();
  assert.deepEqual(trace, ["run", "step", "run"]); assert.equal(k.table[1], 14);
  // the JS core, same shape
  const c0 = new Cell(); let v0 = 0; const jt = [];
  const c = new Constraint("x", () => { c0.track(); jt.push("run"); return v0 * 2; }, () => {});
  c.run(); jsSettle(); jt.length = 0;
  afterSettle(() => { jt.push("step"); v0 = 7; c0.changed(); });
  v0 = 3; c0.changed(); jsSettle();
  assert.deepEqual(jt, trace);
});

console.log(`${passed} passed${process.exitCode ? ", with failures" : ""}`);
